Improve panInside documentation

daverayment opened the pull request:
This is a small fix to resolve issue #7372. It includes an improvement to the documentation to ensure users are aware the padding parameters are `Point` types and it directly links to the `fitBounds` options section which describes each option in more detail.

In addition, there is a new check at the start of the function which sets both X and Y padding correctly if only one value has been supplied; this is a friendlier alternative to raising an error in the console if a `Point` was mistakenly not given.

johnd0e commented:
> this is a friendlier alternative to raising an error

I believe this is error prone approach.
Better either allow and document that officially, or forbid that and throw error.

daverayment commented:
@johnd0e Thanks. I'm now throwing an error if the padding is invalid, and have added a test to exercise this.

johnd0e commented:
Looks good.
But may be we should also enhance docs, introducing dedicated "Padding options" section (instead of [FitBounds options](https://leafletjs.com/reference-1.7.1.html#fitbounds-options))?

daverayment commented:
@johnd0e I agree. Having `fitBounds options` inherit from a new `padding options` is cleaner.

I have made these changes to the documentation in the latest commit, and also added error checking and tests for `topLeftPadding` and `bottomRightPadding` params.

johnd0e commented on src/map/Map.js:
```diff
@@ -511,17 +511,18 @@ export var Map = Evented.extend({
-	// `padding`, `paddingTopLeft` and `paddingTopRight` options
-	// ([Point](#point) types) to fit the display to more restricted bounds,
-	// like the [`padding options for fitBounds`](#fitbounds-options).
+	// `padding`, `paddingTopLeft` and `paddingBottomRight` options
+	// ([Point](#point) types) to fit the display to more restricted bounds.
```
```suggestion
	// padding options to fit the display to more restricted bounds.
```

johnd0e commented:
> Having `fitBounds options` inherit from a new `padding options` is cleaner.

I like it!
But that also makes clear that `fitBounds` also requires similar handling (and tests), right?

daverayment commented:
> But that also makes clear that `fitBounds` also requires similar handling (and tests), right?

Good point! Since `fitBounds` and `flyToBounds` have optional padding, too, I've added a check to the `_getBoundsCenterZoom` function, which they both call.

I've also added tests for both methods to exercise this check. There were no pre-existing tests for `flyToBounds`, so I started a new `describe` section for that in MapSpec.js. I hope that's OK.

One potential concern: this could throw errors for existing projects because we're being stricter with the parameter checking for these 3 functions. (However, affected projects would not have been behaving correctly.) What are your thoughts?

johnd0e commented:
> There were no pre-existing tests for `flyToBounds`, so I started a new `describe` section for that in MapSpec.js. I hope that's OK.

Sure that's ok.
But! This is not enough in general, as `flyToBounds` needs more specific tests.
Perhaps this is out of scope of this PR, but you should at least add some TODO note somewhere.

It's also possible to gather all related tests in one `describe` block.
And maybe we should be more specific about error, checking `.to.throw('Padding must be a point type.');`
<details>

```js
	describe('#padding options', function () {
		beforeEach(function () {
			//...
		});
		var paddingError = 'Padding must be a point type.';

		['padding', 'paddingTopLeft', 'paddingBottomRight'].forEach(function (prop) {
			var paddingOptions = {[prop]: 50};

			it('throws an error with invalid ' + prop, function () {
				expect(function () {
					map.flyToBounds(bounds, paddingOptions);
				}).to.throwError(paddingError);

				expect(function () {
					map.fitBounds//...
				}).to.throwError(paddingError);

				expect(function () {
					map.panInside//...
				}).to.throwError(paddingError);
			});
		});
	});
```
</details>


> One potential concern: this could throw errors for existing projects because we're being stricter with the parameter checking

Yeah, after upgrading maintainers should fix bugs in their code, it's easy because they will know exact cause.

daverayment commented:
Thanks, @johnd0e. I like your approach - it keeps all the checks in one place in MapSpec.js and can easily be expanded upon if other methods have padding options in the future. I've implemented this suggestion and removed the previous tests from elsewhere (including the new `flyToBounds` block). If further work on `flyToBounds` testing is required in the future, I agree that that's probably a separate piece of work.

johnd0e commented on spec/suites/map/MapSpec.js:
```diff
@@ -791,6 +791,36 @@ describe("Map", function () {
 	});
 
+	describe('#padding options types', function () {
+		beforeEach(function () {
```
Perhaps we may use `before`, to execute this block just once before all included tests.

daverayment replied:
I think this needs to be a `beforeEach`, like the other tests, because of the scope of the map object. It doesn't have any measurable performance impact, so I think we should keep it like this to be consistent.

johnd0e commented on spec/suites/map/MapSpec.js:
```diff
@@ -791,6 +791,36 @@ describe("Map", function () {
+
+		['padding', 'paddingTopLeft', 'paddingBottomRight'].forEach(function (prop) {
+			var paddingOptions = {};
+			paddingOptions[prop] = 50; // invalid
```
In fact there is a lot of other invalid values possible, like `[50]` or `"[50]"`. Not sure that we should check over all of them.
But if we not - then better refine the comment to something like "// most common misuse"...

daverayment replied:
The intention is not to test for all possible invalid values, so I'm happy to update the comment.

johnd0e commented on src/map/Map.js:
```diff
@@ -252,6 +252,13 @@ export var Map = Evented.extend({
 
 		options = options || {};
+
+		if ((options.padding && options.padding.length !== 2) ||
```
Invalid value like '50' will pass this test too.

daverayment replied:
This is just to catch cases where a single numeric value is passed in, as it was to deal with the original #7372 issue. Yes, a two-character string will be classed as valid, but I think that's beyond the scope of this simple fix.

johnd0e commented on src/map/Map.js:
```diff
@@ -511,15 +518,20 @@ export var Map = Evented.extend({
 	panInside: function (latlng, options) {
 		options = options || {};
 
+		if ((options.padding && options.padding.length !== 2) ||
```
Another occurrence of same code. Not good...

daverayment replied:
Is the issue that I haven't factored out the error check into a separate fn? I can certainly do that if we decide these checks are enough for now.

johnd0e replied:
> if we decide these checks are enough for now.

I tend to think that current checks definitely not enough.
Why should we care padding options more that others?
And even if we do, why we do not care about similar occurrences:
- https://leafletjs.com/reference-1.7.1.html#marker-autopanpadding
- https://leafletjs.com/reference-1.7.1.html#popup-autopanpaddingtopleft (there are 3)
- etc

So I propose following:
1. Exclude validation code (and corresponding testes) from this PR, leave docs enhancement only (which alone solves most of original #7372 issue).
2. Make another PR adding point object validation code to `toPoint`.

daverayment replied:
The other padding parameters you mention were explicitly documented as taking `Point`s. The `panInside` documentation originally lacked this explicit mention, which contributed to the original misunderstanding by the OP and the #7397 issue being raised. I suspect adding `toPoint` validation may end up catching fewer genuine issues while potentially adding a performance penalty for larger projects.

However, we can open that up for discussion on a separate PR! I'm happy to remove the options checks here and submit just the documentation updates for now.

johnd0e commented on src/map/Map.js:
```diff
@@ -511,15 +518,20 @@ export var Map = Evented.extend({
+			throw new Error('Padding must be a point type.');
+		}
+
 		var paddingTL = toPoint(options.paddingTopLeft || options.padding || [0, 0]),
```
The best solution might be including those checks into `toPoint` function.
What do you think?

johnd0e replied:
As for me - I believe that it would be most clean solution, with much more positive impact, because it is able to catch Point misuse in every place, not only in padding options.
But I have some concerns about (1) compatibility and (2) performance.
1. Perhaps not an issue. Who would want to call `toPoint(50)` intentionally?
2. We need to make sure that `toPoint` is not used in any hot path.

daverayment replied:
I thought about this a few days ago, but I discounted it because I've used `toPoint` in the past on quite large datasets and I was worried the additional checks could affect performance in such cases. I think putting validation on padding options and improving the documentation as we've done here is a good start, and maybe `toPoint` can be revisited in the future.

johnd0e replied:
> I discounted it because I've used `toPoint` in the past on quite large datasets and I was worried the additional checks could affect performance in such cases.

Still, I believe that it is very important - to catch bugs early, so I've added this case to this discussion https://github.com/Leaflet/Leaflet/issues/7128#issuecomment-760765116

johnd0e reviewed (changes requested):
@daverayment 
Could you please resolve merge conflicts?

johnd0e commented:
@daverayment ?
